### PR TITLE
Add Mockgen Version Flag

### PR DESCRIPTION
**Pull request by nguyenfilip**
Add -version flag that shows a version provided by Go modules.

This strategy works well for module aware invocations of `go get` and `go
run` (which means using GO111MODULE=on).

If we ever plan to distribute built binaries, it is necessary to amend
the strategy with build flags.

GOPATH mode versioning is unsupported.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,14 +8,17 @@ contexts too.
...
 ------------
 
-Once you have [installed Go][golang-install], install the `mockgen` tool:
+Once you have [installed Go][golang-install], install the `mockgen` tool.
+
+To get the latest released version use:
 
 ```bash
-go get github.com/golang/mock/mockgen
+GO111MODULE=on go get github.com/golang/mock/mockgen@latest
```
Comment (codyoss) [inline README.md]: These extra options should not be needed with the current version of go right?

File: README.md [hunk 2]
```diff
@@ -8,14 +8,17 @@ contexts too.
...
 ```
 
-_Note: It is recommended to have `GO111MODULE=on` to ensure the correct
-dependencies are used._
+If you use `mockgen` in your CI pipeline, it may be more appropriate to fixate
```
Comment (codyoss) [inline README.md]: Eventually in a wiki we might want to show the tool install pattern used with modules. I can point to an example later.

File: mockgen/version.1.11.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// Copyright 2010 Google Inc.
```
Comment (codyoss) [inline mockgen/version.1.11.go]:
```suggestion
// Copyright 2019 Google LLC
```
The guidance in LLC now.

File: mockgen/version.1.12.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2010 Google Inc.
```
Comment (codyoss) [inline mockgen/version.1.12.go]:
```suggestion
// Copyright 2019 Google LLC
```

File: mockgen/version.1.12.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+import (
+	"fmt"
+	"log"
+	"runtime/debug"
+)
+
+func printVersion() {
+	if bi, exists := debug.ReadBuildInfo(); exists {
+		fmt.Println(bi.Main.Version)
```
Comment (cvgw) [inline mockgen/version.1.12.go]: I haven't seen this pattern before (usually see a const str). Does this return the version metadata of the go module? Does that mean it won't work unless installed via module support? I'm guessing that's what the build checks and error messaging are about.

Comment (codyoss) [inline mockgen/version.1.12.go]: What does this actually print out today if you run it on go1.12+?

File: README.md (hunk 1 above)
Comment (nguyenfilip) [reply README.md]: Actually it is needed because `GO111MODULE` is by default set to `auto`. That means that if you invoke `go get` in a directory outside the `GOPATH` and without go.mod file, then the `go get` executes in  `GOPATH mode`. These are the docs:
```
 If GO111MODULE=auto or is unset, then the go command enables or disables module support based on the current directory. Module support is enabled only when the current directory contains a go.mod file or is below a directory containing a go.mod file.
```
(source: https://golang.org/cmd/go/#hdr-Module_support)

File: README.md (hunk 2 above)
Comment (nguyenfilip) [reply README.md]: ok

File: mockgen/version.1.12.go (hunk 5 above)
Comment (nguyenfilip) [reply mockgen/version.1.12.go]: I am not leaning strongly towards the implementation in this PR, but I tend to think it's the best out of the three options that I have considered:

1. using constant string
2. using build flags
3. using debug.ReadBuildInfo

Each option has pros and cons. 

Problem with constant string is that any time somebody pulls master version X (not the latest version) of mockgen with `go get`, they will always end up with mockgen that includes all the commits since X but the version string is unchanged (the binary would report same version for binaries with different commits, which is exactly what we dont want).

Problem with build flags is that they won't work unless we distribute binaries. And I see that current assumption is that people use `go get` to get the mockgen

Problem with debug.ReadBuildInfo is that it relies on using modules. 

I lean towards using debug.ReadBuildInfo. I think it should be the best practice for the users of `mockgen` to always use predictable version, especially for CI builds. That was the original problem in #250. Thus I think that users should not run `go get` in GOPATH mode and they should use `@` to specify which version of mockgen they want to use. Note that the version info is embedded in the binary, so after `go get` is used to download the binary the clients can copy the `mockgen` out of the GOPATH and `-version` should still work.

As I mentioned in the description of this PR (`the GOPATH mode versioning is unsupported`) this wont work unless installed via module support 

WDYT?

Comment (nguyenfilip) [reply mockgen/version.1.12.go]: As soon as we release 1.3.2, the `-version` flag should report `v1.3.2` if `go get` is invoked with `@latest`. I think there is no way to test this with `mockgen` itself now. But I tested the `debug.ReadBuildInfo` in my other test project. 

One edge case is when somebody invokes the `go get` with `@master`. But it does behave correctly. With `@master` the resulting version should be something like: `v1.3.2-0.20191220092011-827cab0badc9`, in other words if the `go get` is requsted to download an unreleased version, the module support adds commit SHA to the version string to disambiguate the version.

There are multiple situations when `-version` is not going to be available (but I think that is fine):
 - when building/running the mockgen module locally the `-version` is `(devel)`
 - when running with Golang older then 1.12 (as seen in the changes in this PR)

Comment (cvgw) [reply mockgen/version.1.12.go]: >Problem with constant string is that any time somebody pulls master version X (not the latest version) of mockgen with go get, they will always end up with mockgen that includes all the commits since X but the version string is unchanged (the binary would report same version for binaries with different commits, which is exactly what we dont want).

I've always used a constant string in projects but I haven't thought about the UX of pulling master in that case. I don't think I have a strong opinion and I don't disagree with the point you've made.

I don't think there is anything wrong with making module support a requirement for the version command to work; as long as we fall back to some other behavior when module support is not enabled.

Comment (cvgw) [reply mockgen/version.1.12.go]: It looks like there is some nice fallback behavior using the build tags for different versions of go :👍

File: mockgen/version.1.11.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+package main
+
+import (
+	"log"
+)
+
+func printVersion() {
+	log.Printf("No version information is available for Mockgen compiled with " +
+		"version 1.11")
```
Comment (codyoss) [inline mockgen/version.1.11.go]: nit: < 1.12

File: README.md (hunk 1 above)
Comment (codyoss) [inline README.md]: As a part of our next release I think we should start uploading artifacts as well. This will allow more people to get the desired behavior out of version. We can use a recent contribution to help with this: https://github.com/golang/mock/pull/229

File: mockgen/version.1.12.go (hunk 5 above)
Comment (nguyenfilip) [reply mockgen/version.1.12.go]: @cvgw wrt build tags. Are you referring to the `// +build go1.12` tag that has been in this PR?

File: mockgen/version.1.11.go (hunk 6 above)
Comment (nguyenfilip) [reply mockgen/version.1.11.go]: done, thanks!

File: README.md (hunk 1 above)
Comment (nguyenfilip) [reply README.md]: Cool! Yes if we build binaries ourselves and we build it with 1.12+ version of golang, then the built binaries should support the -version flag as implemented here.

File: mockgen/version.1.12.go (hunk 5 above)
Comment (nguyenfilip) [reply mockgen/version.1.12.go]: > I've always used a constant string in projects but I haven't thought about the UX of pulling master in that case. I don't think I have a strong opinion and I don't disagree with the point you've made.

Yeah I agree. Constant string is a good solution if you build and distribute the binary. But! :-) If we did build and distribute the binary, it think there is even better solution than version string. The better solution is to use `-ldflags` with `go build` command. That solution allows you to set a value to global property in your golang program during build time.

Comment (cvgw) [reply mockgen/version.1.12.go]: > @cvgw wrt build tags. Are you referring to the `// +build go1.12` tag that has been in this PR?

yes

>The better solution is to use `-ldflags` with `go build` command. That solution allows you to set a value to global property in your golang program during build time.

I realized that's what we do during compile on another project I help on https://github.com/GoogleContainerTools/kaniko/blob/master/Makefile#L30
